File: src/schemas/macro.ts
import { z } from "zod";
import { datasetEnvelope, FreshnessStatus, utcDateTime } from "./envelope";

export const MacroUnit = z.enum(["pct", "bps", "index", "usd", "ratio", "level"]);

export const MacroIndicator = z
  .object({
    key: z.string().min(1),
    label: z.string().min(1),
    value: z.number().finite().nullable(),
    previous: z.number().finite().nullable(),
    change_1m: z.number().finite().nullable(),
    unit: MacroUnit,
    source: z.string().min(1),
    updated_at: utcDateTime.nullable(),
    status: FreshnessStatus,
  })
  .strict();

export const FedWatchRateProb = z
  .object({
    target_rate: z.number().finite(),
    probability: z.number().finite().min(0).max(1),
    change_1d: z.number().finite().nullable(),
  })
  .strict();

export const FedWatchSnapshot = z
  .object({
    meeting_date: utcDateTime.nullable(),
    effective_rate: z.number().finite(),
    implied_rate: z.number().finite(),
    probabilities: z.array(FedWatchRateProb),
    inferred_action: z.enum(["hold", "hike", "cut", "insufficient_data"]).nullable(),
    change_1d: z.record(z.number().finite()).nullable(),
    status: z.enum(["accumulating", "ready"]),
  })
  .strict();

export const MacroDataset = z
  .object({
    rates: z.array(MacroIndicator).default([]),
    credit: z.array(MacroIndicator).default([]),
    inflation: z.array(MacroIndicator).default([]),
    labor: z.array(MacroIndicator).default([]),
    liquidity: z.array(MacroIndicator).default([]),
    fx: z.array(MacroIndicator).default([]),
    fedwatch: FedWatchSnapshot.nullable(),
  })
  .strict();

export const MacroEnvelope = datasetEnvelope(MacroDataset);

export type MacroUnit = z.infer<typeof MacroUnit>;
export type MacroIndicator = z.infer<typeof MacroIndicator>;
export type FedWatchRateProb = z.infer<typeof FedWatchRateProb>;
export type FedWatchSnapshot = z.infer<typeof FedWatchSnapshot>;
export type MacroDataset = z.infer<typeof MacroDataset>;
export type MacroEnvelope = z.infer<typeof MacroEnvelope>;

File: pipeline/schemas/macro.py
"""Macro dataset contract (architecture §3.2 indicator mapping / review §3.1 FRED cornerstone).

payload structure: MacroIndicator lists grouped by business domain + FedWatch snapshot.
Filled by the T03 MacroCollector; this module only defines the contract.
"""

from __future__ import annotations

from typing import Literal

from pydantic import Field

from .envelope import BaseEnvelope, ContractModel, FreshnessStatus, UTCDateTime

MacroUnit = Literal["pct", "bps", "index", "usd", "ratio", "level"]


class MacroIndicator(ContractModel):
    """A single macro indicator (raw numeric storage, architecture §8.3)."""

    key: str = Field(min_length=1)
    label: str = Field(min_length=1)
    value: float | None = None
    previous: float | None = None
    change_1m: float | None = None
    unit: MacroUnit = "level"
    source: str = Field(min_length=1)
    updated_at: UTCDateTime | None = None
    status: FreshnessStatus = "fresh"


class FedWatchRateProb(ContractModel):
    """Probability of a target rate bucket (0-1)."""

    target_rate: float
    probability: float = Field(ge=0.0, le=1.0)
    change_1d: float | None = None


class FedWatchSnapshot(ContractModel):
    """CME FedWatch probability self-computed snapshot (architecture §1.6 frozen methodology)."""

    meeting_date: UTCDateTime | None = None
    effective_rate: float
    implied_rate: float
    probabilities: list[FedWatchRateProb] = Field(default_factory=list)
    inferred_action: Literal["hold", "hike", "cut", "insufficient_data"] | None = None
    change_1d: dict[str, float] | None = None
    status: Literal["accumulating", "ready"] = "accumulating"
    # Free settlement history is only ~5 trading days (review P0-1): "change vs a week ago" is
    # unavailable until the system has run for 7 full days; the frontend must show the
    # insufficient-data status instead of 0/empty.


class MacroDataset(ContractModel):
    """Macro dataset payload."""

    rates: list[MacroIndicator] = Field(default_factory=list)
    credit: list[MacroIndicator] = Field(default_factory=list)
    inflation: list[MacroIndicator] = Field(default_factory=list)
    labor: list[MacroIndicator] = Field(default_factory=list)
    liquidity: list[MacroIndicator] = Field(default_factory=list)
    fx: list[MacroIndicator] = Field(default_factory=list)
    fedwatch: FedWatchSnapshot | None = None


class MacroEnvelope(BaseEnvelope):
    """macro.json envelope (strongly typed payload)."""

    payload: MacroDataset
